pkg/xds/server/resource_warming_forcer.go

```diff
@@ -0,0 +1,124 @@
+//
+// The solution
+//   The easiest solution would be to just set a new version of endpoints also when cluster changes (in pkg/xds/server/SnapshotReconciler) . The problem is that go-control-plane does not support resource ordering
+//   * https://github.com/envoyproxy/go-control-plane/issues/59
+//   * https://github.com/envoyproxy/go-control-plane/issues/218
+//   * https://github.com/envoyproxy/go-control-plane/issues/235
+//   Therefore even if we were to set a new version of EDS + CDS on one snapshot, there is no guarantee that EDS will be delivered after CDS.
+//
+//   The alternative solution is this based on a callbacks.
+//   Nonce is a sequence indicator of a sent DiscoveryResponse on a stream. We use ADS therefore every single DiscoveryResponse regardless of a type is sent with incremented nonce.
+//   Typical sequence of CDS + EDS looks like this:
+//   1) Envoy sends DiscoveryRequest [type=CDS, version="", nonce=""] // ask for the clusters
+//   2) Kuma sends DiscoveryResponse [type=CDS, version="UUID-1", nonce="1"] // response with clusters
+//   3) Envoy sends DiscoveryRequest [type=EDS, version="", nonce=""] // ask for the endpoints
+//   4) Envoy sends DiscoveryRequest [type=CDS, version="UUID-1", nonce="1"] // confirmations that it received clusters (ACK)
+//   5) Kuma sends DiscoveryResponse [type=EDS, version="UUID-2", nonce="2"] // response with endpoints
+//   6) Envoy sends DiscoveryRequest [type=EDS, version="UUID-2", nonce="2"] // confirmations that it received endpoints (ACK)
+//
+//   Then if we send a Cluster update (continuing the flow above)
+//   7) Kuma sends DiscoveryResponse [type=CDS, version="UUID-2", nonce="3"] // response with cluster update
+//   8) Envoy sends DiscoveryRequest [type=CDS, version="UUID-2", nonce="3"] // ACK
+//   9) Envoy sends DiscoveryRequest [type=EDS, version="UUID-2", nonce="2"] // Envoy sends a request which looks like the second ACK for the previous endpoints
+//
+//   Updated Cluster is now in warming state until we send DiscoveryResponse with EDS.
+//   We could sent the same DiscoveryResponse again: DiscoveryRequest [type=EDS, version="UUID-2", nonce="2"], but there is no API in go-control-plane to do it.
+//   Instead we set a new version of the Endpoints to force a new EDS exchange:
+//   10) Kuma sends DiscoveryResponse [type=EDS, version="UUID-3", nonce="3"] // triggered because we set snapshot with a new version
+//   11) Envoy sends DiscoveryRequest [type=EDS, version="UUID-3", nonce="3"] // ACK
+//   After this exchange, cluster is now out of the warming state.
+//
+// The same problem is with Listeners and Routes (change of the Listener that uses RDS requires RDS DiscoveryResponse), but since we don't use RDS now, the implementation is for EDS only.
+// More reading of how Envoy is trying to solve it https://github.com/envoyproxy/envoy/issues/13009
```

Comment:
Quite an explanation here. Is this a new problem or we just port an old thing to the new world?

Reply:
Old thing ported as a callback. It was coded here pkg/xds/server/server.go (look for peekResponse, scheduleResponse)